pkg/segment/query/querystatus.go

```diff
@@ -176,18 +190,65 @@ func StartQuery(qid uint64, async bool, cleanupCallback func()) (*RunningQuerySt
 		return nil, fmt.Errorf("qid has already been started")
 	}
 
-	if len(allRunningQueries) >= MAX_RUNNING_QUERIES {
+	stateChan := make(chan *QueryStateChanData, numStates)
+
+	runningState := &RunningQueryState{
+		startTime:         time.Now(),
+		StateChan:         stateChan,
+		cleanupCallback:   cleanupCallback,
+		rqsLock:           &sync.Mutex{},
+		isAsync:           async,
+		timeoutCancelFunc: nil,
+	}
+
+	select {
+	case waitingQueries <- WaitStateData{qid, runningState}:
+	default:
 		return nil, putils.TeeErrorf("StartQuery: qid=%v cannot be started, Max number of running queries reached", qid)
 	}
 
-	var stateChan chan *QueryStateChanData
-	if async {
-		stateChan = make(chan *QueryStateChanData, numStates)
-		stateChan <- &QueryStateChanData{StateName: RUNNING}
+	return runningState, nil
+}
+
+// Removes reference to qid. If qid does not exist this is a noop
+func DeleteQuery(qid uint64) {
+	LogGlobalSearchErrors(qid)
+	arqMapLock.Lock()
+	rQuery, ok := allRunningQueries[qid]
+	if ok {
+		delete(allRunningQueries, qid)
+	}
+	arqMapLock.Unlock()
+
+	if ok && !rQuery.isCancelled {
+		rQuery.timeoutCancelFunc()
+
+		if rQuery.cleanupCallback != nil {
+			rQuery.cleanupCallback()
+		}
+	}
+}
+
+func canRunQuery() bool {
+	activeQueries := GetActiveQueryCount()
+	return (activeQueries < MAX_RUNNING_QUERIES && memory.FreeMemory()+QUERY_MEM <= config.GetTotalMemoryAvailable())
+}
+
+func PullQueriesToRun() {
+	for {
+		if canRunQuery() {
+			select {
+			case wsData := <-waitingQueries:
+				runQuery(wsData)
+			default:
+			}
+		}
+		time.Sleep(10 * time.Millisecond)
```

Comment:
nit: make this a constant at the top of the file

Reply:
addressed in the latest commit.